datajunction-server/datajunction_server/sql/dag.py

```diff
@@ -102,16 +103,32 @@ async def get_downstream_nodes(
 
     paths = dag.union_all(next_layer)
 
-    final_select = select(Node)
+    # Calculate the maximum depth for each node
+    max_depths = (
+        select(paths.c.node_id, func.max(paths.c.depth).label("max_depth"))
+        .group_by(paths.c.node_id)
+        .cte("max_depths")
+    )
+
+    # Select nodes with the maximum depth
+    final_select = (
+        select(Node, max_depths.c.max_depth)
+        .join(max_depths, max_depths.c.node_id == Node.id)
+    )
+
     if not include_deactivated:
         final_select = final_select.where(is_(Node.deactivated_at, None))
+
+    # Add depth filter
+    if depth != -1:
```

Comment:
Tiny change but cleaner imo:
```suggestion
    if depth > -1:
```

Reply:
good idea, changed!